weave/feedback.py

```diff
@@ -0,0 +1,230 @@
+
+
+class Feedbacks(AbstractRichContainer[tsi.Feedback]):
+    """A collection of Feedback objects with utilities."""
+
+    show_refs: bool
+
+    def __init__(
+        self, show_refs: bool, feedbacks: Optional[Iterable[tsi.Feedback]] = None
+    ) -> None:
+        super().__init__("Feedback", feedbacks)
+        self.show_refs = show_refs
+
+    def refs(self) -> Refs:
+        """Return the unique refs associated with these feedbacks."""
+        uris = list(set(feedback.weave_ref for feedback in self.items))
+        return Refs(uris)
+
+    def _add_table_columns(self, table: Table) -> None:
+        if self.show_refs:
+            table.add_column("Ref", overflow="fold")
+        table.add_column("Type", justify="center")
+        table.add_column("Feedback", overflow="fold")
+        table.add_column("Created")
+        table.add_column("ID", overflow="fold")
+        table.add_column("Creator")
+
+    def _item_to_row(self, item: tsi.Feedback) -> list:
+        feedback = item
+
+        typ = feedback.feedback_type
+        display_type = typ
+        if typ == "wandb.reaction.1":
+            display_type = "reaction"
+            if util.is_notebook():
+                # TODO: Emojis mess up table alignment in Jupyter 😢
+                #       maybe there is something else we could do here?
+                content = feedback.payload["alias"]
+            else:
+                content = feedback.payload["emoji"]
+        elif typ == "wandb.note.1":
+            display_type = "note"
+            content = feedback.payload["note"]
+        else:
+            content = json.dumps(feedback.payload, indent=2)
+
+        # TODO: Prettier relative time display?
+        created_at = str(feedback.created_at.replace(tzinfo=None))
+
+        creator = feedback.wb_user_id
+        if feedback.creator is not None:
+            creator = f"{feedback.creator} ({creator})"
+
+        row = [
+            display_type,
+            content,
+            created_at,
+            feedback.id,
+            creator,
+        ]
+        if self.show_refs:
+            row.insert(0, feedback.weave_ref)
+        return row
+
+
+class FeedbackQuery:
+    """Lazy-loading object for fetching feedback from the server."""
+
+    entity: str
+    project: str
+
+    show_refs: bool
+    _query: tsi.Query
+    offset: Optional[int]
+    limit: Optional[int]
+
+    feedbacks: Optional[Feedbacks]
+
+    def __init__(
+        self,
+        entity: str,
+        project: str,
+        query: Query,
+        offset: Optional[int] = None,
+        limit: Optional[int] = None,
+        show_refs: bool = False,
+    ):
+        self.client = graph_client_context.require_graph_client()
+        self.entity = entity
+        self.project = project
+
+        self.show_refs = show_refs
+        self._query = query
+        self.offset = offset
+        self.limit = limit
+
+        self.feedbacks = None
+
+    def __iter__(self) -> Iterator[tsi.Feedback]:
```

Comment:
any particular reason you excluded the len/index operations?

Reply:
Nope, just missed during refactor.